core/object/object.cpp

```diff
@@ -982,30 +981,34 @@ Error Object::emit_signalp(const StringName &p_name, const Variant **p_args, int
 
 		const Variant **args = p_args;
 		int argc = p_argcount;
+		Callable ca = c.callable;
 
 		if (c.flags & CONNECT_APPEND_SOURCE_OBJECT) {
-			argc = p_argcount + 1;
-			append_args.resize(argc);
-
-			for (int j = 0; j < p_argcount; j++) {
-				append_args.write[j] = p_args[j];
-			}
-
 			if (unlikely(appended_self.get_type() == Variant::NIL)) {
 				appended_self = this;
 			}
-			append_args.write[p_argcount] = &appended_self;
 
-			args = (const Variant **)append_args.ptr();
+			if (ca.is_custom()) {
```

Comment:
I think your original approach was not bad, you can do:
```c++
const Variant ** append_args;
Variant appended_self;
// then later on
if (unlikely(appended_self.get_type() == Variant::NIL)) {
   append_args = (const Variant**)alloca(sizeof(Variant*) * (p_argcount+1));
   for (..) copy all arg pointers
   appended_self=this
   append_args[p_argcount]=&appended_self;
```
And then just pass this to the Callable, it will ensure the self is appended after the _call_ arguments (not the bind) and should be very performant.

Reply:
Actually, what I've been worried about is this case:
![0](https://user-images.githubusercontent.com/30386067/185814378-affc5ed0-08c4-418c-bb74-dfb643643ab5.png)

https://github.com/godotengine/godot/blob/0c5f254956f0115e363ce08045dd178dc30b54f8/core/variant/callable_bind.cpp#L167-L175

This feature should not conflict with **unbind**. But if appending to the end, they are conflicting.

Reply:
I think unbind just happens on another level and you should ignore it. Its up to the user.